### DISCUSSION TITLE
External sensor support

### DISCUSSION THREAD

**Discussion by dhuzjak** [omerfaruk-aran/esphome_samsung_hvac_bus#216] [Ideas]
Hi,

few days ago I've added this project to my Samsung 4-way cassette unit which didn't come with WiFi. Everything works like a charm. So I was experimenting with heating mode, and as now I'm collecting data from AC, I've noticed that device is turning off and on every 4 minutes. Its internal temperature sensor reaches target temperature very quickly, while room temperature takes longer than that.

Here is the graph
<img width="1129" alt="SamsungAC_heating_graph" src="https://github.com/user-attachments/assets/bbca23b5-1f38-4425-be09-914c1a75a425">

Top graph shows target temperature (red), room temperature (yellow), internal sensor temperature (blue). On below graph we see outdoor power which shows that outdoor unit is turning on and off about every 4 minutes.

I was wondering if it would be possible to add external sensor to this project so AC would adjust its output based on sensor in the room instead of internal sensor. E.g. have one sensor that would push data to MQTT topic, then esphome_samsung_hvac device would read from this topic and send it to AC. Or something that would emulate wired external controller. 

Is something like that possible?

Thanks

**Comment by omerfaruk-aran** [comment]
Hello,

First of all, could you clarify where the automation is running? Is the automatic control of your HVAC system being managed on Home Assistant or is it being controlled by the internal mechanisms of the Samsung device?

If the automation is running on Home Assistant, it is possible to modify the operation of the air conditioner using temperature data from an external sensor. In this scenario, you can update the automation rules on Home Assistant to change the device's operation or modes based on external sensor data.

However, if the Samsung device is managing its own internal control mechanisms, it is not practically possible to interfere with its internal sensor readings. Devices like these typically rely on their own sensors for internal control and do not accept external data to override those readings. Therefore, it would not be possible to change the device's operation mode using an external sensor in this case.

To provide more assistance and suggestions, could you share the complete details of your HVAC system? Information about the product models you are using, additional software (such as Home Assistant or any other automation systems), and other relevant details would help us provide better guidance.

**Comment by dhuzjak** [comment]
There is no automation, I didn't say there is. This is pure data logging from regular AC operation in heating mode. My indoor unit is AJ026TNNDKG if it helps.

Samsung has wired room controller (like MWR-WG00JN) which I assume has temperature sensor. You can connect that controller and I would expect that you can configure AC unit to take readings from wired room controller sensor instead of internal one.

So idea is that ESP32 would emulate external controller like MWR-WG00JN and send room temperature via bus. ESP32 would get room temperature data either physically from wired sensor or wireless via network. Then AC would compensate power output based on external sensor data instead of internal.

**Reply by omerfaruk-aran** [reply]
Hello,

Thank you for providing more information. However, the solution you proposed falls outside the current scope of my project. Emulating Samsung’s wired room controller (like MWR-WG00JN) to send temperature data directly to the AC unit would be quite complex and require a detailed effort. At the moment, I do not have the time to dedicate to this task.

It might be more beneficial to find another developer who can allocate time and resources to this project. If I have time to work on this in the future, I will reconsider the possibility.

Thank you for your interest and understanding.

**Reply by bzumik1** [reply]
I am also interested in this functionality. Not expert in this field but I thought it will be as simple as sending correct message on the bus to simulate readings from the wired thermostat.

**Comment by nicolaerosia** [comment]
I’m also interested in this.

The room controller is on F3/F4? If yes, i did not find any documentation on that protocol, it seems those 2 wires carry both data and power

**Comment by dhuzjak** [comment]
@bzumik1 That was also my initial idea. That some message can be sent on the bus that would simulate external sensor.
Looking at this [docs](https://wiki.myehs.eu/wiki/NASA_Protocol) there is even message like
<small><small>
0x4204 | VAR_IN_?? | NASA_MODIFIED_CURRENT_TEMP | Temperature
-- | -- | -- | --

</small></small>
And I thought maybe this can be sent to override internal sensor reading. But I agree that can go in wrong direction because AC has its own math to provide correct output and this is calibrated for internal sensor.

Regarding external controller, there is some information in [this](https://github.com/DannyDeGaspari/Samsung-HVAC-buscontrol) project, but its non-NASA protocol.

**Comment by tech2urdoor** [comment]
I have a wired controller and I had to configure the controller to get it's temperature reading from the controller rather than the sensor on the intake. I did this when I had the official wifi kit installed and when I replaced the kit with the atom it still gets that same information so I didn't have to make a temperature offset. Don't know if you can configure your controller the same way but it would be outside the scope of this project. You would need to look into programming your actual wired controller first.

**Reply by dhuzjak** [reply]
This comment missed the point. Where did I say I have wired controller? I don't have intention to spend additional money for wired controller, neither I don't want to run wires on my wall just to have external controller.
I suggested that I would like to see functionality to take readings from external sensor and send it to AC. Since AC already has support for external controller, it means its not impossible.

Project name is "esphome_samsung_hvac_bus" so I dont see how its outside of scope to send some values to the bus. But if maintainer has no time or resources to work on it, I can accept that. I don't really need it that bad, but I think this would be a nice feature.

**Reply by tech2urdoor** [reply]
I'm sorry I guess I'm just a bit confused by your wording. If you had your AC connected to home assistant you would be able to automate the on and off functions of your AC using any temperature input sensor, located anywhere you liked.  This integration allows you to control the AC. Also as you stated yourself, you can get a wired controller connected to your AC and this integration could  get its data from that. It sounds like you are asking him to rewrite this code to incorporate an external sensor which can all be done through home assistant automations. I'm really sorry if I'm just misunderstanding but that's just how I'm reading your request....

**Reply by nicolaerosia** [reply]
he *probably* wants to be able to use the water law to control the output temperature based on interior and exterior temperature. On / Off controller is so 1990 :)

**Reply by tech2urdoor** [reply]
Maybe I'm just an idiot then but I still don't understand. Is law an acronym for something? Could he not write an automation to adjust the target temperature of the AC based on any sensor reading he wanted through home assistant? It doesn't have to be a simple on and off. I just feel like you can't really expect someone to code a complete virtual wired controller. That would take a huge amount of time for what seems like little benefit. I'm not trying to start an argument here or anything. I'm just genuinely confused and don't want anyone wasting time coding all that...as I said earlier. Maybe I'm just missing the whole point and I'm just stupid.

**Reply by nicolaerosia** [reply]
you’re not an idiot, it can be done. Water law is the name for the algorithm which computes the leaving temperature.
From my understanding, he wants to be able to send that message to the AC unit, as he assumes that would tell the AC what is the interior temperature, just like a wired controller would do, and the AC algo would just work

**Reply by nicolaerosia** [reply]
probably it makes sense that there would be a service call exposed to home assistant that allows you to push any payload you wish

**Comment by tech2urdoor** [comment]
I guess I'm only just seeing it all from my little bubble. I replaced my official Samsung wifi kit with an esp chip because they stopped supporting their app which I used to control my AC originally. The wifi kit didn't have any sensors in it at all. It just read data from the bus and was able to send basic commands like, on/off, mode changes, and target temperatures. I'm truly amazed by how much extra functionality is already provided like outdoor sensor temperature and power values that I could have only dreamed of getting from an official Samsung device. It was going to cost me $380 AUD for an upgraded wifi kit (if I could even find a non-nasa version) which would have connected to smart things. They would have inevitably stopped supporting that in the future as well. The atom chip and rs485 adapter cost me $30 all up and thanks to the hard work of programmers and this community I'm now free from the cloud and no-one can take it away from me. I just feel like some people's expectations are a bit too high. This is all just my perspective though. I'm Australian, we don't get snow here and I don't even really understand HVAC. It's just AC here. Please forgive my ignorance.

**Comment by omerfaruk-aran** [comment]
Okay, no problem, everyone, calm down :). I am closing this discussion, but I will look at this issue in detail later when I have free time. Please be patient and wait. I thank everyone for their contributions.
